### ISSUE TITLE
Remember commit message for current branch when closing / re-opening Lazygit

### ISSUE THREAD

**Issue by lougreenwood** [jesseduffield/lazygit#2720]
**Is your feature request related to a problem? Please describe.**
I use Lazygit in nvim using `toggleterm`.

Often I try to commit and I get a pre-commit hook error from some linter, so I need to close the lazygit float term, fix the error and re-open it. When I re-open lazy got my old commit message is gone, so I need to re-type it again.

**Describe the solution you'd like**
If Lazygit could remember the commit message for the current branch until that commit is successful it would improve my workflow. The MacOS Git GUI app, `Fork` has a similar feature which saves the commit message, this is where my inspiration came from.

I would guess that Lazygit could forget the commit message if all changes are discarded. 

**Describe alternatives you've considered**
setting up nvim-remote so I don't need to close the toggle term, but this seems like a work around for this solution, I imagine that my use case isn't the only one which would benefit from lazygit remembering the commit message.

**Comment by mark2185** [comment]
Does `toggleterm` _have to_ close the running program? Can't it just hide it and show it again?

Also, does that happen in regular usage of `lazygit`, i.e. outside of `nvim`?
 I'm not sure if an error because of `pre-commit hooks` clears the commit message, and I'm too lazy to set it up.

Having `lazygit` remember a commit message _across restarts_ isn't really needed nor wanted.

**Comment by stefanhaller** [comment]
> I'm not sure if an error because of pre-commit hooks clears the commit message,

Yes it does, see https://github.com/jesseduffield/lazygit/issues/2679. Pretty annoying, actually.

**Comment by dagadbm** [comment]
I was just about to suggest this actually.

I have some precommit hooks for eslint and sometimes I forget dangling code which is caught on commit.
This closes lazygit and effectively I end up loosing the commit message and I need to retype it again.

It would be nice if we could press up arrow and it would give you the last commit message you wrote (no matter if the commit was succesfull or not)

**Comment by robclancy** [comment]
The commit description being cleared is extremely frustrating to the point I won't use this to commit anymore incase I close my draft to check something.

**Comment by jesseduffield** [comment]
We could store the commit message in the state.yml file. This is the same place that we store the list of recent repos. I bet this would be easy to implement. We'd add a field for commit messages [here](https://github.com/jesseduffield/lazygit/blob/57bb1aa69853de458faf886f077dfca9250dc28d/pkg/config/app_config.go#L315) like so:
```go
// AppState stores data between runs of the app like when the last update check
// was performed and which other repos have been checked out
type AppState struct {
	LastUpdateCheck     int64
	RecentRepos         []string
	StartupPopupVersion int

	// these are for custom commands typed in directly, not for custom commands in the lazygit config
	CustomCommandsHistory []string
	HideCommandLog        bool
	// Map from the path of the worktree to the state of the worktree
	Worktrees             map[string]WorktreeState
}

type WorktreeState struct {
	// Draft commit messages exist against worktrees, not repos, because in one repo
	// there may be multiple worktrees, and each worktree may have a different draft
	// commit message
	DraftCommitMessage string
}
```

Then in https://github.com/jesseduffield/lazygit/blob/57bb1aa69853de458faf886f077dfca9250dc28d/pkg/gui/controllers/helpers/commits_helper.go#L114, rather than call `GetPreservedMessage()` and `SetPreservedMessage()` on the commit message context, we would have the helper itself define those methods and have them read/write to that state file e.g. like:
```go
worktreeState, ok := self.c.GetAppState().Worktrees[self.c.WorktreePath()]
if !ok {
  worktreeState = config.WorktreeState{}
  self.c.GetAppState().Worktrees[self.c.WorktreePath()] = worktreeState
}
worktreeState.DraftCommitMessage = draftCommitMessage
err := self.c.SaveAppState()
```

There are a few things we'd need to do like have a method for getting the current worktree path (my worktrees [PR](https://github.com/jesseduffield/lazygit/pull/2147) isn't in yet). But otherwise it sounds fairly straightforward.

Something worth thinking about, though is whether we want to introduce a pattern for per-repo (or worktree) state/config that lives inside the repo. This would prevent issues when repos are moved, because there are no hard-coded paths involved. I think this is worth thinking about

**Comment by robclancy** [comment]
I did a long commit message the other day knowing to not close it because it decides I want to type it out twice. But little did I know that scrolling the diff behind the box will also close it and goodbye commit message.

**Comment by jesseduffield** [comment]
@robclancy it should be storing both the summary and the description. Are you saying that it currently only stores the summary? If so could you raise a separate issue for that?

**Comment by robclancy** [comment]
Yes, and I realize this is probably the wrong issue. The summary is always saved fine and the description is always lost.

**Comment by jesseduffield** [comment]
Good to know!

**Comment by divaltor** [comment]
What's the status of issue?

**Comment by elvinmahmudov** [comment]
What is the status of this ? It has been more than a year and still no updated? @jesseduffield

**Comment by jesseduffield** [comment]
The status is that it's still a great idea, we should totally do it, it's totally doable, but it hasn't been a priority.

In response to some comments I made earlier:

> There are a few things we'd need to do like have a method for getting the current worktree path (my worktrees https://github.com/jesseduffield/lazygit/pull/2147 isn't in yet). But otherwise it sounds fairly straightforward.

This is now implemented.

> Something worth thinking about, though is whether we want to introduce a pattern for per-repo (or worktree) state/config that lives inside the repo (i.e. inside the .git folder). This would prevent issues when repos are moved, because there are no hard-coded paths involved. I think this is worth thinking about

Just storing the state in the global state file seems fine to me. Yes, if you move the repo, you'll lose your mid-way commit message, but that is a very rare edge case, and we do per-repo state files as a second step if we want.

If anybody wants to take a stab at this, I'm happy to provide guidance. (tagging @karimkhaleel @AzraelSec @jwhitley in case one of you was looking for a new issue!).

If you yourself would like to work on this @elvinmahmudov let me know.

**Comment by AzraelSec** [comment]
I was looking for something to work on - I'd be up to give this a shot if it's fine for you @jesseduffield

**Comment by stefanhaller** [comment]
> Just storing the state in the global state file seems fine to me. 

Another option could be to store the message in `.git/COMMIT_EDITMSG`. This is the file that [git itself uses to store the message](https://www.git-scm.com/docs/git-commit#_files), so it would give us a bit of interoperability with git cli; for example, you could start typing the message in lazygit, then cancel, and type shift-C to proceed in the editor, and your message would show up there (I think). Or vice versa.

The only complication is that we'd have to strip comments from the file (lines beginning with `#`) when reading from it, in case the commit was started in git.

**Comment by jesseduffield** [comment]
@stefanhaller  Ah yes, I like that a lot

@AzraelSec sounds good! I've assigned to you

**Comment by jesseduffield** [comment]
hey @AzraelSec are you still interested in adding this feature?

**Comment by AzraelSec** [comment]
Hey @jesseduffield, that was my plan (I actually already had a working branch locally), but I really had no time in the past few months. I should be able to work on this in the following weeks, but if there's anyone else who's available right now to put some effort into implementing this, I'm happy to step back.

**Comment by jesseduffield** [comment]
No worries. I'll leave you assigned in the meantime, but if anybody else wants to pick this up, lemme know

**Comment by AzraelSec** [comment]
@stefanhaller @jesseduffield I tried to use the `.git/COMMIT_EDITMSG` file to store the in-flight commit message as suggested in a previous comment, but any following `git commit` invocation causes that file to be deleted, as the documentation states:

```
[...] any commit message [...] will be overwritten by the next invocation of git commit
```

Starting to type the commit message in lazygit and then continuing using the external editor would not be possible anyway. Do you think it still makes sense to handle a different file instead of using the one that hosts the global state?

**Comment by jesseduffield** [comment]
Huh that's surprising: it begs the question of why it's in a file in the first place rather than just being stored in memory.

What happens if you do `git commit --edit --file .git/COMMIT_EDITMSG`? Like in https://stackoverflow.com/questions/20580611/how-to-reuse-pending-comments-after-git-commit-fail

The benefit of that that file over a random file chosen by lazygit is that if a git commit is started outside lazygit and it fails, the message so-far will be stored in that file.

**Comment by stefanhaller** [comment]
Ok, the behavior of the `COMMIT_EDITMSG` file is not what I thought it was. I thought git cli would 1) remove the file after a successful commit, 2) leave it in place after a failed commit, and 3) reuse the file automatically when it exists. Of these, only 2) is true, the other two are not. This makes the whole approach very questionable. Especially 1) is a problem: if you make a successful commit with git cli, and then open lazygit and make another commit there, you will get the earlier commit's message in lazygit; that's not what we want.

So it looks like we should store it privately on lazygit's side then (probably in state.yml).

**Comment by jesseduffield** [comment]
Ah yes, point 1 is a show-stopper. Okay, using our own private file makes sense then

**Comment by AzraelSec** [comment]
@jesseduffield @stefanhaller - this issue can be closed since the feature has been merged into master with #4191. I wrongly removed the "closes" statement from that PR's description.

**Comment by jesseduffield** [comment]
Nice one @AzraelSec
